### PR TITLE
[Parser] Support class definitions and properties definitions

### PR DESCRIPTION
**Pull request by shumbo**
## Overview

This PR implements the new strategy to compile Scenic class definitions and property definitions.

An overview of the strategy can be found in [my note](https://shun-notes.notion.site/How-to-compile-class-definitions-482061588cd14ec091c51e74d7eaba8b).

## Change in Grammar

Several changes were made to the grammar to generate AST nodes that represent property definitions.

In Scenic, properties can be added to the class using the following syntax.

```python
class MyScenicClass(Object):
    # without attributes
    property: value
    # with attributes
    property[additive]: value
```

Although the code above is syntactically valid in Python, it has a different semantic, and we want to generate different types of AST nodes.

This PR adds `PropertyDef` for this purpose. `PropertyDef` has three fields: `property` for the name of the property, `attributes` for meta attributes and `value` for the default value of the property.

Note that `PropertyDef` can only be at the top level of the class definition body. For example, the following code would not produce `PropertyDef` and instead generate Python's `AnnAssign`.

```python
class MyScenicClass(Object):
    def method():
        property: value # -> AnnAssign to the variable `property`
        property[additive]: value # -> AnnAssign to `property` subscript `additive`
```

To achieve this, I added `scenic_class_def_block`, `scenic_class_statements` and `scenic_class_statement` to the grammar. They are almost identical to the normal `block` `statements` `statement`, but in addition to all statements, it allows `scenic_class_property_stmt` with the highest precedence. By having `scenic_class_def_block` for class body, property definitions are parsed as `PropertyDef`.

Only the two literal `additive` and `dynamic` can be inside the square brackets (with a comma to separate the two if need to specify both). If a different literal is specified, that raises a syntax error.

## Compiling `ClassDef` `PropertyDef`

With this PR, we are moving away from passing default values using annotations. Instead of annotations, the compiler creates a special dictionary `_scenic_properties` that stores all properties and corresponding values for a class.

As stated, `PropertyDef` can only be in a body of `ClassDef`. On `visit_ClassDef`, we extract all property definitions and turn them into a single dictionary declaration.

For example, 

```python
class MyScenicClass(Object):
    property1: value1
    property2: value2
```

would be compiled to

```python
class MyScenicClass(Object):
    _scenic_properties = {
        "property1": PropertyDefault(set(), set(), lambda self: value1),
        "property2": PropertyDefault(set(), set(), lambda self: value2),
    }
```

Because we do this transformation in `ClassDef`, `visit_PropertyDef` should not be called. If the function is called, it will raise an assertion error.

`visit_ClassDef` also adds `Object` for the base class if no base classes were specified.

## Reading property values

`object_types.py` was changed to read property values from `_scenic_properties` instead of `__annotations__`.

The file has definitions of the built-in types (`Object`, `Point`, `OrientedPoint`). They were rewritten to define properties using `_scenic_properties`.

Finally, `__init__subclass__` on `Constructible` now asserts that no more than one Scenic class was given as the base class to enforce single inheritance of Scenic classes.

## `new` veneer function

The `new` keyword now asserts that the given class is a Scenic class (i.e., the subclass of `Constructible`).

The `new` keyword is compiled to a call to the `new` veneer function. It performs the check, raises an exception if the check fails, and returns an instance.

### CODE REVIEW

File: tests/syntax/test_classes.py [hunk 1]
```diff
@@ -72,13 +61,13 @@ class Bar(Foo):
 def test_isinstance_issubclass():
     scenario = compileScenic("""
         class Foo: pass
-        ego = Foo
+        ego = new Foo
         if isinstance(ego, Foo):
```
Comment (shumbo) [inline tests/syntax/test_classes.py:65]: needs #62 to correctly compile `ego` and this case doesn't pass at the moment.

File: tests/syntax/test_classes.py [hunk 2]
```diff
@@ -46,9 +35,9 @@ class Foo:\n
 def test_property_simple():
     scenario = compileScenic("""
         class Foo:
-            position: 3 @ 9
+            position: (3, 9)
```
Comment (shumbo) [inline tests/syntax/test_classes.py:38]: `@` needs #56, so changing it to tuples. Can revert the change once #56 is merged.

File: src/scenic/syntax/scenic.gram [hunk 3]
```diff
@@ -701,6 +701,37 @@ class_def_raw[ast.ClassDef]:
...
+            property=a.string,
+            attributes=b if b is not None else [],
+            value=c,
+            LOCATIONS,
+        )
+     }
+
+# fail if `NAME [ <expr> ]` pattern is found at top level of class definition and
+# <expr> is neither `additive` nor `dynamic`
+scenic_class_property_attribute: &&(
```
Comment (shumbo) [inline src/scenic/syntax/scenic.gram:730]: [`&&`](https://we-like-parsers.github.io/pegen/grammar.html#e-6) is an operator that fails immediately if the expression after that fails to parse.

If what follows after `[` is not `additive` or `dynamic`, it raises a syntax error.
Screen Shot 2022-07-14 at 9 08 17

Comment (shumbo) [reply src/scenic/syntax/scenic.gram:730]: I first wrote a rule like the following:

```
scenic_class_property_attribute:
    | "additive" { s.Additive(LOCATIONS) }
    | "dynamic" { s.Dynamic(LOCATIONS) }
    | !NAME
```

However, if you put a non-Name expression inside the square brackets (like quoted strings or numbers), this rule would fail but will be parsed as a normal annotated assignment.

We can list possible tokens and fail all of them, but I think it would be better to use `&&` here.

File: tests/syntax/test_classes.py [hunk 4]
```diff
@@ -5,39 +5,28 @@
...
 
 def test_old_constructor_statement():
-    compileScenic("""
-        constructor Foo:
-            blah: 19 @ -3
-        ego = Foo with blah 12
-    """)
+    with pytest.raises(SyntaxError):
+        compileScenic("""
+            constructor Foo:
```
Comment (shumbo) [inline tests/syntax/test_classes.py:12]: the new parser does not support the old `constructor` keyword. Use of the keyword would result in a syntax error.

File: src/scenic/syntax/scenic.gram (hunk 3 above)
Comment (dfremont) [reply src/scenic/syntax/scenic.gram:730]: Agreed.

**Review by dfremont — APPROVED** [review]
Looks good to me!
